api/consumers.py

```diff
@@ -183,10 +184,48 @@ def update_quiz_end_time(self):
 
     async def end_quiz(self):
         if await self.update_quiz_end_time():
-            await self.send(text_data=json.dumps({"type": "quiz_ended"}))
+            grades = await self.fetch_grades()
+            await self.send(
+                text_data=json.dumps(
+                    {
+                        "type": "quiz_ended",
+                        "grades": grades,
+                    }
+                )
+            )
         else:
             print("Failed to end the quiz; session not found.")
 
+    @database_sync_to_async
+    def fetch_grades(self):
+        try:
+            session = QuizSession.objects.get(code=self.code)
+        except QuizSession.DoesNotExist:
+            return {}
+
+        students = session.students.all()
+        grade_buckets = defaultdict(list)
+
+        for student in students:
+            responses = student.responses.all()
+            total_responses = responses.count()
+            correct_responses = responses.filter(is_correct=True).count()
+
+            # Calculate percentage
+            if total_responses > 0:
+                percentage = (correct_responses / total_responses) * 100
+            else:
+                percentage = 0.0
+
+            # Round to two decimal places
+            percentage = round(percentage, 2)
+            percentage_key = f"{percentage}"
+
+            # Append the student's username to the appropriate bucket
+            grade_buckets[percentage_key].append(student.username)
+
+        return grade_buckets
```

Comment:
There might be an issue with this logic now that I'm taking a second look; `total_responses = responses.count()` may not be the number of questions in the quiz; if a student does not submit an answer, there won't be a record in that table.

Total responses should be retrieved from `api_quiz_session_question`, and it should ignore skipped questions, when you get `correct_responses` you should also ignore the questions that were skipped (using `skipped` attribute in `api_quiz_session_question` and not `skipped_question` in `api_user_response`)

Also, how have you tested this? Is there an easy way for the PR reviewer to test this?